graph/src/components/store.rs

```diff
@@ -1716,3 +1743,47 @@ impl LfuCache<EntityKey, Option<Entity>> {
         }
     }
 }
+
+/// Determines which columns should be selected in a table.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum ColumnNames {
+    /// Select all columns. Equivalent to a `"SELECT *"`.
+    All,
+    /// Individual column names to be selected.
+    Select(BTreeSet<String>),
+}
+
+impl ColumnNames {
+    pub fn insert(&mut self, column_name: &str) {
+        match self {
+            ColumnNames::All => {
+                let mut set = BTreeSet::new();
+                set.insert(column_name.to_string());
+                *self = ColumnNames::Select(set)
+            }
+            ColumnNames::Select(set) => {
+                set.insert(column_name.to_string());
+            }
+        }
+    }
+
+    pub fn update(&mut self, field: &q::Field) {
+        // ignore "meta" field names
+        if field.name.starts_with("__") {
+            return;
+        }
+        self.insert(&field.name)
+    }
+
+    pub fn extend(&mut self, other: Self) {
+        use ColumnNames::*;
+        match (self, other) {
+            (All, All) => {}
+            (self_ @ All, other @ Select(_)) => *self_ = other,
+            (Select(_), All) => {
+                unreachable!()
```

Comment:
Is it safe to leave this here? And even then, shouldn't this be `*self = All`?

Reply:
Agree.

It may sound counterintuitive, but since `self` is more specific, I believe it should keep its original value in that case.
However, this might lead to "under-fetching," which undesirable. 

In conclusion, I think that using `All` in this case is a safe assumption.

Reply:
I agree, this `unreachable!()` should go out.

As the `All` variant encompasses all the table's columns, this might be counterintuitive, but I believe `self` should keep the original value in this case since it is more specific.

On the other hand, that might lead to unforeseen "under-fetching" scenarios, even though the `All` variant is currently used only in cases where we are uncertain of which columns to select and all of our tests, for that matter, are passing. 

I don't mean to digress, but we could consider the creation of a new variant to represent that state:
```enum ColumnNames {
    All,
   Select(BTreeSet<String>),
   Unspecified,
}```

Reply:
In theory I agree that `*self = All` seems correct. But since this is never used anywhere I'd keep the `unreachable!()`, though I'm fine with either.

@tilacog I see what you mean that `All` is a bit ambiguous, bu that's fine imo.

Reply:
My concern with this is that it's very unintuitive that the operation is not symmetric, i.e. that `a.extend(b)` and `b.extend(a)` do not yield the same result. It makes sense for how this is currently used, but if a future code change changes the order in which entries are traversed, it could lead to incorrect query execution (I _think_ the GraphQL layer would blow up if we do not fetch all required columns) But I now think that if we leave the `unreachable!` there, we should be safe since things will blow up loudly if we change the order in which attribute names are traversed and collected.